README.md
# nifti-hs
This library will supply pure transformations and processing for the NIFTI file format.


ChangeLog.md
# Changelog for nifti-hs

## Unreleased changes


src/Lib.hs
module Lib
    ( copyNIFTI
    ) where

import System.IO
import qualified Data.ByteString.Lazy as BL
import Data.Binary.Get
import Data.Word

deserialiseHeader :: Get (Word32, Word32, Word32)
deserialiseHeader = do
  alen <- getWord32be
  plen <- getWord32be
  chksum <- getWord32be
  return (alen, plen, chksum)

getHandles :: IO (Handle, Handle)
getHandles = do
    -- TODO: make this parse cmd line args instead of prompting twice
    in_filename <- getLine
    out_filename <- getLine
    in_handle <- openBinaryFile in_filename ReadMode
    out_handle <- openBinaryFile out_filename WriteMode
    return (in_handle, out_handle)

copyNIFTI :: IO ()
copyNIFTI = do
    (inh, outh) <- getHandles
    rawData <- hGetContents inh -- read content lazily
    hPutStr outh rawData -- write content
    -- TODO: do processing here
    --close handles
    hClose inh
    hClose outh

src/Header.hs
module Header
    ( getNifti1HeaderLE
    ) where

import System.IO
import qualified Data.ByteString.Lazy as BL
import Data.Binary.Get
import Data.Word
import Data.Text

-- array types used in header
type Dim = (Word16, Word16, Word16, Word16, Word16, Word16, Word16, Word16)
type PixDim = (Float, Float, Float, Float, Float, Float, Float, Float)

-- header itself
data Nifti1Header = Nifti1Header
    { sizeof_hdr        :: Word32      -- MUST be 3416
    , dim_info          :: Word16       -- MRI slice ordering
    , dim               :: Dim              -- array dimensions
    , intent_p1         :: Float            -- 1st intent parameter
    , intent_p2         :: Float            -- 2nd intent parameter
    , intent_p3         :: Float            -- 3rd intent parameter
    , intent_code       :: Word16       -- NIFTIINTENT code
    , data_type         :: Word16       -- Defines data type 
    , bitpix            :: Word16       -- number of bits/voxel
    , slice_start       :: Word16       -- first slice index
    , pixdim            :: PixDim           -- grid spacing
    , vox_offset        :: Float            -- offset into .nii file
    , scl_slope         :: Float            -- scaling: slope
    , scl_inter         :: Float            -- data scaling: offset
    , slice_end         :: Word16       -- Last slice index
    , slice_code        :: Word16       -- Slice timing order
    , xyzt_units        :: Word16       -- Units of pixdim[1..4]
    , cal_max           :: Float            -- Max display intensity
    , cal_min           :: Float            -- Min display intensity
    , slice_duration    :: Float            -- Time for 1 slice
    , toffset           :: Float            -- Time axis shift
    , descrip           :: Text             -- Any text description, total size is 160 bytes
    , aux_file          :: Text             -- auxiliary filename, total size is 24 bytes
    , qform_code        :: Word16       -- NIFTIXFORM code
    , sform_code        :: Word16       -- NIFTIXFORM code
    , quatern_b         :: Float            -- Quaternion b param
    , quatern_c         :: Float            -- Quaternion c param
    , quatern_d         :: Float            -- Quaternion d param
    , qoffset_x         :: Float            -- Quaternion x shift
    , qoffset_y         :: Float            -- Quaternion y shift
    , qoffset_z         :: Float            -- Quaternion z shift
    , srow_x            :: Float            -- 1st row affine transform TODO: as array of floats
    , srow_y            :: Float            -- 2nd row affine transform TODO: as array of floats
    , srow_z            :: Float            -- 3rd row affine transform TODO: as array of floats
    , intent_name       :: Text           -- Name or meaning of data
    , magic             :: Text           -- MUST be "ni1\0" or "n+1\0"
    }

deserialiseDim :: Get Dim
deserialiseDim = do
  d0 <- getWord16le
  d1 <- getWord16le
  d2 <- getWord16le
  d3 <- getWord16le
  d4 <- getWord16le
  d5 <- getWord16le
  d6 <- getWord16le
  d7 <- getWord16le
  return (d0, d1, d2, d3, d4, d5, d6, d7)

deserialisePixDim :: Get PixDim
deserialisePixDim = do
  d0 <- getFloatle
  d1 <- getFloatle
  d2 <- getFloatle
  d3 <- getFloatle
  d4 <- getFloatle
  d5 <- getFloatle
  d6 <- getFloatle
  d7 <- getFloatle
  return (d0, d1, d2, d3, d4, d5, d6, d7)

getNifti1HeaderLE :: Get Nifti1Header
-- Nifti1 Little Endian
getNifti1HeaderLE = do
    sizeof_hdr <- getWord32le
    skip 35 -- These 35 bytes of the header are unused
    dim_info <- getWord16le
    dim <- deserialiseDim
    intent_p1 <- getFloatle
    intent_p2 <- getFloatle
    intent_p3 <- getFloatle
    intent_code <- getWord16le
    data_type <- getWord16le
    bitpix <- getWord16le
    slice_start <- getWord16le
    pixdim <- deserialisePixDim
    vox_offset <- getFloatle
    scl_slope <- getFloatle
    scl_inter <- getFloatle
    slice_end <- getWord16le
    slice_code <- getWord16le
    xyzt_units <- getWord16le
    cal_max <- getFloatle
    cal_min <- getFloatle
    slice_duration <- getFloatle
    toffset <- getFloatle
    --descrip <- getByteString 160 -- read exactly 160 bytes
    skip 160 -- skip this for now because i don't know how to handle it
    -- aux_file <- getByteString 40 -- read exactly 40 bytes
    skip 40 -- skip for now
    qform_code <- getWord16le
    sform_code <- getWord16le
    quatern_b <- getFloatle
    quatern_c <- getFloatle
    quatern_d <- getFloatle
    qoffset_x <- getFloatle
    qoffset_y <- getFloatle
    qoffset_z <- getFloatle
    srow_x <- getFloatle
    srow_y <- getFloatle
    srow_z <- getFloatle
    -- intent_name <- getByteString 16 -- read exactly 16 bytes
    skip 16 -- skip for now
    -- magic <- getByteString 4 -- read exactly 4 bytes
    skip 4  -- skip for now
    return $! Nifti1Header
      { sizeof_hdr = sizeof_hdr
      , dim_info = dim_info
      , dim = dim
      , intent_p1 = intent_p1
      , intent_p2 = intent_p2
      , intent_p3 = intent_p3
      , intent_code = intent_code
      , data_type = data_type
      , bitpix = bitpix
      , slice_start = slice_start
      , pixdim = pixdim
      , vox_offset = vox_offset
      , scl_slope = scl_slope
      , scl_inter = scl_inter
      , slice_end = slice_end
      , slice_code = slice_code
      , xyzt_units = xyzt_units
      , cal_max = cal_max
      , cal_min = cal_min
      , slice_duration = slice_duration
      , toffset = toffset
      -- , descrip = descrip
      -- , aux_file = aux_file
      , qform_code = qform_code
      , sform_code = sform_code
      , quatern_b = quatern_b
      , quatern_c = quatern_c
      , quatern_d = quatern_d
      , qoffset_x = qoffset_x
      , qoffset_y = qoffset_y
      , qoffset_z = qoffset_z
      , srow_x = srow_x
      , srow_y = srow_y
      , srow_z = srow_z
      -- , intent_name = intent_name
      -- , magic = magic
      }

{-
incrementalExample :: BL.ByteString -> [Nifti1Header]
incrementalExample input0 = go decoder input0
  where
    decoder = runGetIncremental getNifti1HeaderLE
    go :: Decoder Nifti1Header -> BL.ByteString -> [Nifti1Header]
    go (Done leftover _consumed hdr) input =
      hdr: go decoder (BL.chunk leftover input)
    go (Partial k) input                     =
      go (k . takeHeadChunk $ input) (dropHeadChunk input)
    go (Fail _leftover _consumed msg) _input =
      error msg

takeHeadChunk :: BL.ByteString -> Maybe BS.ByteString
takeHeadChunk lbs =
  case lbs of
    (BL.Chunk bs _) -> Just bs
    _ -> Nothing

dropHeadChunk :: BL.ByteString -> BL.ByteString
dropHeadChunk lbs =
  case lbs of
    (BL.Chunk _ lbs') -> lbs'
    _ -> BL.Empty

-}

src/Endianness.hs
module Endianness where

import Data.Word
import Data.Int
import qualified Data.Binary.Get as G

data Endianness
  = Little
  | Big
  | Host

getWord8 :: G.Get Word8
getWord8 = G.getWord8

getWord16 :: Endianness -> G.Get Word16
getWord16 Little = G.getWord16le
getWord16 Big = G.getWord16be
getWord16 Host = G.getWord16host

getWord32 :: Endianness -> G.Get Word32
getWord32 Little = G.getWord32le
getWord32 Big = G.getWord32be
getWord32 Host = G.getWord32host

getWord64 :: Endianness -> G.Get Word64
getWord64 Little = G.getWord64le
getWord64 Big = G.getWord64be
getWord64 Host = G.getWord64host

getInt32 :: Endianness -> G.Get Int32
getInt32 Little = G.getInt32le
getInt32 Big = G.getInt32be
getInt32 Host = G.getInt32host

getFloat :: Endianness -> G.Get Float
getFloat Little = G.getFloatle
getFloat Big = G.getFloatbe
getFloat Host = G.getFloathost


src/nifti_header.hs
-- array types used in header
type Dim = (Data.Word8, Data.Word8, Data.Word8, Data.Word8, Data.Word8, Data.Word8, Data.Word8, Data.Word8,)
type PixDim = (Float, Float, Float, Float, Float, Float, Float, Float,)

-- header itself
data Nifti1Header = Nifti1Header
    { sizeof_hdr        :: Data.Word32      -- MUST be 348
    , dim_info          :: Data.Word8       -- MRI slice ordering
    , dim               :: Dim              -- Data array dimensions
    , intent_p1         :: Float            -- 1st intent parameter
    , intent_p2         :: Float            -- 2nd intent parameter
    , intent_p3         :: Float            -- 3rd intent parameter
    , intent_code       :: Data.Word8       -- NIFTIINTENT code
    , data_type         :: Data.Word8       -- Defines data type 
    , bitpix            :: Data.Word8       -- number of bits/voxel
    , slice_start       :: Data.Word8       -- first slice index
    , pixdim            :: PixDim           -- grid spacing
    , vox_offset        :: Float            -- offset into .nii file
    , scl_slope         :: Float            -- Data scaling: slope
    , scl_inter         :: Float            -- data scaling: offset
    , slice_end         :: Data.Word8       -- Last slice index
    , slice_code        :: Data.Word8       -- Slice timing order
    , xyzt_units        :: Data.Word8       -- Units of pixdim[1..4]
    , cal_max           :: Float            -- Max display intensity
    , cal_min           :: Float            -- Min display intensity
    , slice_duration    :: Float            -- Time for 1 slice
    , toffset           :: Float            -- Time axis shift
    , descrip           :: String           -- Any text description, total size is 80 bytes
    , aux_file          :: String           -- auxiliary filename, total size is 24 bytes
    , qform_code        :: Data.Word8       -- NIFTIXFORM code
    , sform_code        :: Data.Word8       -- NIFTIXFORM code
    , quatern_b         :: Float            -- Quaternion b param
    , quatern_c         :: Float            -- Quaternion c param
    , quatern_d         :: Float            -- Quaternion d param
    , qoffset_x         :: Float            -- Quaternion x shift
    , qoffset_y         :: Float            -- Quaternion y shift
    , qoffset_z         :: Float            -- Quaternion z shift
    , srow_x            :: Float            -- 1st row affine transform TODO: as array of floats
    , srow_y            :: Float            -- 2nd row affine transform TODO: as array of floats
    , srow_z            :: Float            -- 3rd row affine transform TODO: as array of floats
    , intent_name       :: String           -- Name or meaning of data
    , magic             :: String           -- MUST be "ni1\0" or "n+1\0"
    }

getNifti1HeaderLE :: Get Nifti1Header
-- Nifti1 Little Endian
getNifti1HeaderLE = do
    sizeof_hdr <- getWord32le
    skip 35 -- These 35 bytes of the header are unused
    dim_info <- getWord8le
    dim <- getWord16le
    intent_p1 <- getFloatle
    intent_p2 <- getFloatle
    intent_p3 <- getFloatle
    intent_code <- getWord8le
    data_type <- getWord8le
    bitpix <- getWord8le
    slice_start <- getWord8le
    pixdim <- getFloatle
    vox_offset <- getFloatle
    scl_slope <- getFloatle
    scl_inter <- getFloatle
    slice_end <- getWord8le
    slice_code <- getWord8le
    xyzt_units <- getWord8le
    cal_max <- getFloatle
    cal_min <- getFloatle
    slice_duration <- getFloatle
    toffset <- getFloatle
    descrip <- getByteString 80 -- read exactly 80 bytes
    aux_file <- getByteString 40 -- read exactly 40 bytes
    qform_code <- getWord8le
    sform_code <- getWord8le
    quatern_b <- getFloatle
    quatern_c <- getFloatle
    quatern_d <- getFloatle
    qoffset_x <- getFloatle
    qoffset_y <- getFloatle
    qoffset_z <- getFloatle
    srow_x <- getFloatle
    srow_y <- getFloatle
    srow_z <- getFloatle
    intent_name <- getByteString 16 -- read exactly 16 bytes
    magic <- getByteString 4 -- read exactly 4 bytes
    return $! Nifti1Header sizeof_hdr dim_info dim intent_p1 intent_p2 intent_p3 intent_code data_type bitpix slice_start pixdim vox_offset scl_slope scl_inter slice_end slice_code xyzt_units cal_max cal_min slice_duration toffset descrip aux_file qform_code sform_code quatern_b quatern_c quatern_d qoffset_x qoffset_y qoffset_z srow_x srow_y srow_z intent_name magic

incrementalExample :: BL.ByteString -> [Nifti1Header]
incrementalExample input0 = go decoder input0
  where
    decoder = runGetIncremental getNifti1HeaderLE
    go :: Decoder Nifti1Header -> BL.ByteString -> [Nifti1Header]
    go (Done leftover _consumed hdr) input =
      hdr: go decoder (BL.chunk leftover input)
    go (Partial k) input                     =
      go (k . takeHeadChunk $ input) (dropHeadChunk input)
    go (Fail _leftover _consumed msg) _input =
      error msg

takeHeadChunk :: BL.ByteString -> Maybe BS.ByteString
takeHeadChunk lbs =
  case lbs of
    (BL.Chunk bs _) -> Just bs
    _ -> Nothing

dropHeadChunk :: BL.ByteString -> BL.ByteString
dropHeadChunk lbs =
  case lbs of
    (BL.Chunk _ lbs') -> lbs'
    _ -> BL.Empty

src/NiftiHeader.hs
module NiftiHeader where

import GHC.TypeLits
import Data.Kind
import Data.Word
import Data.Proxy
import Data.Type.Equality
import Control.Monad (when)

import GHC.TypeLits.Compare
import Data.Vector.Unboxed.Sized
import Data.Binary.Get (Get)
import qualified Data.ByteString.Lazy as BL
import qualified Data.Text as T

import qualified Endianness as E

type DimSize (n::Nat) = (1 <= n, n <= 7)

data DimSizeWitness (n::Nat) where
  DimSizeWitness :: ((1 <=? n) :~: True) -> ((n <=? 7) :~: True) -> DimSizeWitness n

isDimSize :: (KnownNat n) => Proxy n -> Maybe (DimSizeWitness n)
isDimSize p = case (isLE p1 p, isLE p p7) of
                (Just Refl, Just Refl) -> Just $ DimSizeWitness Refl Refl
                _ -> Nothing
  where p1 = (Proxy :: Proxy 1)
        p7 = (Proxy :: Proxy 7)

data SomeSize (t :: Nat -> Type) where
  SomeSize :: (DimSize n) => t n -> SomeSize t

-- array types used in header
data Dim (n::Nat) where
  Dim :: (DimSize n) => Vector n Word8 -> Dim n

type SomeDim = SomeSize Dim

data PixDim (n::Nat) where
  PixDim :: (DimSize n) => Vector n Float -> PixDim n

-- header itself
data Nifti1Header (n::Nat) = DimSize n => Nifti1Header
    { sizeof_hdr        :: Word32      -- MUST be 348
    , dim_info          :: Word8       -- MRI slice ordering
    , dim               :: Dim n            -- Data array dimensions
    , intent_p1         :: Float            -- 1st intent parameter
    , intent_p2         :: Float            -- 2nd intent parameter
    , intent_p3         :: Float            -- 3rd intent parameter
    , intent_code       :: Word8       -- NIFTIINTENT code
    , data_type         :: Word8       -- Defines data type
    , bitpix            :: Word8       -- number of bits/voxel
    , slice_start       :: Word8       -- first slice index
    , pixdim            :: PixDim n         -- grid spacing
    , vox_offset        :: Float            -- offset into .nii file
    , scl_slope         :: Float            -- Data scaling: slope
    , scl_inter         :: Float            -- data scaling: offset
    , slice_end         :: Word8       -- Last slice index
    , slice_code        :: Word8       -- Slice timing order
    , xyzt_units        :: Word8       -- Units of pixdim[1..4]
    , cal_max           :: Float            -- Max display intensity
    , cal_min           :: Float            -- Min display intensity
    , slice_duration    :: Float            -- Time for 1 slice
    , toffset           :: Float            -- Time axis shift
    , descrip           :: T.Text           -- Any text description, total size is 80 bytes
    , aux_file          :: T.Text           -- auxiliary filename, total size is 24 bytes
    , qform_code        :: Word8       -- NIFTIXFORM code
    , sform_code        :: Word8       -- NIFTIXFORM code
    , quatern_b         :: Float            -- Quaternion b param
    , quatern_c         :: Float            -- Quaternion c param
    , quatern_d         :: Float            -- Quaternion d param
    , qoffset_x         :: Float            -- Quaternion x shift
    , qoffset_y         :: Float            -- Quaternion y shift
    , qoffset_z         :: Float            -- Quaternion z shift
    , srow_x            :: Float            -- 1st row affine transform TODO: as array of floats
    , srow_y            :: Float            -- 2nd row affine transform TODO: as array of floats
    , srow_z            :: Float            -- 3rd row affine transform TODO: as array of floats
    , intent_name       :: T.Text           -- Name or meaning of data
    , magic             :: T.Text           -- MUST be "ni1\0" or "n+1\0"
    }

type SomeNifti1Header = SomeSize Nifti1Header

getDim :: E.Endianness -> Get SomeDim
getDim e = do
  length <- E.getWord8
  Just (SomeNat lengthProxy) <- someNatVal length
  case isDimSize lengthProxy of
    Nothing -> fail "Invalid length"
    Just (DimSizeWitness Refl Refl)-> do
      vec <- replicateM' lengthProxy $ E.getWord8 e
      return $ SomeSize $ Dim vec

getPixDim :: (KnownNat n, DimSize n) => Proxy n -> E.Endianness -> Get (PixDim n)
getPixDim p e = do
  let length = natVal p + 1
  vec <- replicateM' length $ E.getFloat e
  return vec

determineNifti1HeaderEndianness :: Get E.Endianness
determineNifti1HeaderEndianness = do
  isLittle <- isEndianness E.Little
  if isLittle
    then return E.Little
    else do
      isBig <- isEndianness E.Big
      if isBig
        then return E.Big
        else fail "sizeof_hdr field invalid"
  where
    isEndianness :: E.Endianness -> Get Bool
    isEndianness e = lookAhead $ do
      size <- E.getInt32 e
      return $ size == 348

getNifti1HeaderE :: E.Endianness -> Get SomeNifti1Header
getNifti1HeaderE e = do
  sizeof_hdr <- E.getWord32 endianness
  when (sizeof_hdr /= 348) $ fail "Invalid sizeof_hdr"

  skip 35 -- These 35 bytes of the header are unused

  dim_info <- E.getWord16 e
  SomeSize (dim :: Dim n) <- getDim e
  intent_p1 <- E.getFloat e
  intent_p2 <- E.getFloat e
  intent_p3 <- E.getFloat e
  intent_code <- E.getWord16 e
  data_type <- E.getWord16 e
  bitpix <- E.getWord16 e
  slice_start <- E.getWord16 e
  pixdim <- getPixDim (Proxy @ n) e
  vox_offset <- E.getFloat e
  scl_slope <- E.getFloat e
  scl_inter <- E.getFloat e
  slice_end <- E.getFloat e
  slice_code <- E.getWord16 e
  xyzt_units <- E.getWord16 e
  cal_max <- E.getFloat e
  cal_min <- E.getFloat e
  slice_duration <- E.getFloat e
  toffset <- E.getFloat e
  --descrip <- getByteString 160 -- read exactly 160 bytes
  let descrip = pack "descrip"
  skip 160 -- skip this for now because i don't know how to handbe it
  -- aux_file <- getByteString 40 -- read exactly 40 bytes
  let aux_file = pack "aux_file"
  skip 40 -- skip for now
  qform_code <- E.getWord16 e
  sform_code <- E.getWord16 e
  quatern_b <- E.getFloat e
  quatern_c <- E.getFloat e
  quatern_d <- E.getFloat e
  qoffset_x <- E.getFloat e
  qoffset_y <- E.getFloat e
  qoffset_z <- E.getFloat e
  srow_x <- E.getFloat e
  srow_y <- E.getFloat e
  srow_z <- E.getFloat e
  -- intent_name <- getByteString 16 -- read exactly 16 bytes
  let intent_name = pack "intent_name"
  skip 16 -- skip for now
  -- magic <- getByteString 4 -- read exactly 4 bytes
  let magic = pack "magic"
  skip 4  -- skip for now
  return $! Nifti1Header
    { sizeof_hdr = sizeof_hdr
    , dim_info = dim_info
    , dim = dim
    , intent_p1 = intent_p1
    , intent_p2 = intent_p2
    , intent_p3 = intent_p3
    , intent_code = intent_code
    , data_type = data_type
    , bitpix = bitpix
    , slice_start = slice_start
    , pixdim = pixdim
    , vox_offset = vox_offset
    , scl_slope = scl_slope
    , scl_inter = scl_inter
    , slice_end = slice_end
    , slice_code = slice_code
    , xyzt_units = xyzt_units
    , cal_max = cal_max
    , cal_min = cal_min
    , slice_duration = slice_duration
    , toffset = toffset
    , descrip = descrip
    , aux_file = aux_file
    , qform_code = qform_code
    , sform_code = sform_code
    , quatern_b = quatern_b
    , quatern_c = quatern_c
    , quatern_d = quatern_d
    , qoffset_x = qoffset_x
    , qoffset_y = qoffset_y
    , qoffset_z = qoffset_z
    , srow_x = srow_x
    , srow_y = srow_y
    , srow_z = srow_z
    , intent_name = intent_name
    , magic = magic
    }

getNifti1Header :: Get SomeNifti1Header
getNifti1Header =
  determineNifti1HeaderEndianness >>= getNifti1HeaderE

decodeNifti1Header :: (DimSize n) => BL.ByteString -> (BL.ByteString, Nifti1Header n)
decodeNifti1Header bs = (unconsumed, hdr)
  where (unconsumed, _, hdr) = runGetOrFail getNifti1HeaderLE bs
